WIP: Requirement lists

hikinggrass opened the pull request.

tmolitor-stud-tu committed:
Implement requirement lists in framework core and js wrapper

Signed-off-by: Thilo Molitor <thilo@eightysoft.de>

tmolitor-stud-tu committed:
Fix js errorhandling

Signed-off-by: Thilo Molitor <thilo@eightysoft.de>

caller commented:
@tmolitor-stud-tu  you last commit wasn't signed, so we cant merge this... can you fix that?

tmolitor-stud-tu commented:
It is still WIP (the C++ code generator stuff is still missing), so nothing to merge yet

tmolitor-stud-tu committed:
Implement C++ requirement lists in framework core

Signed-off-by: Thilo Molitor <thilo@eightysoft.de>

tmolitor-stud-tu committed:
Use defaults from meta schemes

Signed-off-by: Thilo Molitor <thilo@eightysoft.de>

tmolitor-stud-tu committed:
Modernize framework error handling

Signed-off-by: Thilo Molitor <thilo@eightysoft.de>

tmolitor-stud-tu committed:
Cleanup manifest meta schema

Signed-off-by: Thilo Molitor <thilo@eightysoft.de>

tmolitor-stud-tu committed:
Don't allow default values for vars and cmd arguments/results

Signed-off-by: Thilo Molitor <thilo@eightysoft.de>

tmolitor-stud-tu committed:
Make sure interface definitions get default values from meta schema

Signed-off-by: Thilo Molitor <thilo@eightysoft.de>

tmolitor-stud-tu committed:
Don't handle rejections, terminate with a backtrace

When using the on_reject callback, all js exceptions thrown in async
functions get silenced. Even worse: there is no way to get the exception
text or js backtrace that led to the exception.
This makes sure we get a proper exception logged and the module
terminated.

Signed-off-by: Thilo Molitor <thilo@eightysoft.de>

a-w50 commented on everestjs/everestjs.cpp:
```diff
@@ -93,8 +93,7 @@ static Napi::Value setup_cmd_handler(const std::string& impl_id, const std::stri
                 },
-                [&result](const Napi::Value& retval, bool) {
-                    // FIXME (aw): we could check the bool for failure ...
+                [&result](const Napi::Value& retval) {
```
I'm not quite sure if this is the right way to go.  Because the on_reject isn't set now anymore, we have no chance to set a custom rejection handler.  I guess now we will only receive an exception, which we can't directly identify, i.e. from which async callback it came from - this would have been possible if we registered custom logic in the on_reject function.
So why we don't check the bool here? If it is true, we know that we're in the on_reject branch and retval should be our error object, which we can handle (that's why there was the FIXME comment) or just throw it right now.

tmolitor-stud-tu replied:
no, to the contrary: we now have a complete js backtrace pointing us directly to the line of code that caused the exception.

tmolitor-stud-tu replied:
Second: why catch it and rethrow it manually when we could leave it to the nodejs to do exactly that?

a-w50 replied:
But where would/could you catch this exception now?  If you want to have the backtrace, you could have just written
if (bool == true) throw Napi::Error
or something similar.  But I would leave it up to the user to decide on how to deal with rejected promises when using the JsExecContext helper class

hikinggrass commented on include/framework/everest.hpp:
```diff
@@ -89,8 +88,8 @@ class Everest {
-    void subscribe_var(const std::string& requirement_id, const std::string& var_name, const ValueCallback& callback);
-    void subscribe_var(const std::string& requirement_id, const std::string& var_name, const JsonCallback& callback);
+    void subscribe_var(const std::string& requirement_id, uint64_t index, const std::string& var_name,
+                       const ValueCallback& callback);
```
I think this might be something we should revisit during a cleanup after implementing type-support.

hikinggrass commented on lib/config.cpp:
```diff
@@ -363,9 +370,28 @@ json Config::load_interface_file(const std::string& intf_name) {
 
         json interface_json = json::parse(intf_file);
 
+        // this subschema can not use allOf with the draft-07 schema because that will cause our validator to
+        // add all draft-07 default values which never validate (the {"not": true} default contradicts everything)
+        // --> validating against draft-07 will be done in an extra step below
```
It refers to a (slightly surprising) behavior of the validator that would populate the patch with these draft-7 default values. Which in turn would lead to some very wrong non-validatable output  that we encountered during testing. Might be useful for future reference, but maybe we should clarify the intent a bit more.

a-w50 replied:
which default values, if I have a non-required integer property and will validate against draft7 with patch, will this lead to the property been set to zero then, or what?

hikinggrass replied:
I have an example from the **enable command** of the **board_support_AC** interface which we found during testing:
In the main branch the validator would create the following output:
```json
{"description":"True: enabled, false: disabled.","type":"boolean"}
```

Without the changes mentioned in the comment we encountered the following output after validating with draft-7 and using the patch from the validation.
```json
{"additionalItems":true,"additionalProperties":true,"contains":true,"description":"True: enabled, false: disabled.","else":true,"if":true,"items":true,"not":true,"propertyNames":true,"readOnly":false,"then":true,"type":"boolean","uniqueItems":false}
```
The way it's implemented now prevents these (obviously wrong) additions to be patched in.

tmolitor-stud-tu replied:
What kai said. Even worse: these additions are contradicting in itself: you never will be able to validate something having `{"not": true}` in the schema no matter what you try to validate.

tmolitor-stud-tu commented on everestjs/utils.hpp:
```diff
@@ -13,52 +13,42 @@ namespace EverestJs {
             throw;                                                                                                     \
-        } catch (Napi::Error & e) {                                                                                    \
+        } catch (Napi::Error& e) {                                                                                     \
             try {                                                                                                      \
                 BOOST_THROW_EXCEPTION(boost::enable_error_info(e)                                                      \
                                       << boost::log::BOOST_LOG_VERSION_NAMESPACE::current_scope());                    \
-            } catch (std::exception & e) {                                                                             \
+            } catch (boost::exception& ex) {                                                                           \
```
It is to get all the boost related error information (the current scope, file, line etc., basically everything you see in the log)

tmolitor-stud-tu commented on lib/config.cpp:
```diff
@@ -733,80 +735,65 @@ void Config::resolve_all_requirements() {
+            for (uint64_t connection_num = 0; connection_num < connections.size(); connection_num++) {
+                auto& connection = connections[connection_num];
+                std::string connection_module_id = connection["module_id"];
+                if (!this->main.contains(connection["module_id"])) {
```
just a copy/paste hiccup, may as well be used, yes.

tmolitor-stud-tu commented on lib/config.cpp:
```diff
@@ -733,80 +735,65 @@ void Config::resolve_all_requirements() {
-                                            " does not provide a implementation for '", connection_impl_id, "'!"));
+                                            printable_identifier(module_id), " not fulfilled: requirement list does",
+                                            "not have an entry count between ", requirement["min_connections"],
+                                            " and ", requirement["max_connections"], "!"));
```
sure

tmolitor-stud-tu commented on lib/config.cpp:
```diff
@@ -733,80 +735,65 @@ void Config::resolve_all_requirements() {
+                    EVLOG_AND_THROW(EVEXCEPTION(EverestConfigError, "Requirement '", requirement_id, "' of module ",
+                                                printable_identifier(module_id), " not fulfilled: module id '",
+                                                connection_module_id,
+                                                "' (configured in connection ", connection_num,
```
sure

tmolitor-stud-tu commented on schemas/interface.json:
```diff
@@ -23,7 +23,8 @@
                             "minLength": 2
-                        }
+                        },
+                        "default": false            // we don't allow default values
```
good question, I guess the current validation process (which includes patching the json with default values) will add defaults if not present in the data, so yes, it will be handled,but I don't know if that is desired behavior.

The better solution would be to remove all entries called `default` when loading the interface definition file, no matter their nesting level.

tmolitor-stud-tu commented on lib/config.cpp:
```diff
@@ -363,9 +370,28 @@ json Config::load_interface_file(const std::string& intf_name) {
         json_validator validator(Config::loader, Config::format_checker);
         validator.set_root_schema(this->_schemas.interface);
-        validator.validate(interface_json);
+        auto patch = validator.validate(interface_json);
+        if (!patch.is_null()) {
+            // extend config entry with default values
+            interface_json = interface_json.patch(patch);
+        }
+
+        // validate every cmd arg/result and var definition against draft-07 schema
+        validator.set_root_schema(draft07);
+        for (auto& var_entry : interface_json["vars"].items()) {
+            validator.validate(var_entry.value());
+        }
+        for (auto& cmd_entry : interface_json["cmds"].items()) {
+            for (auto& arguments_entry : interface_json["cmds"][cmd_entry.key()]["arguments"].items()) {
+                validator.validate(arguments_entry.value());
+            }
+            validator.validate(interface_json["cmds"][cmd_entry.key()]["result"]);
+        }
```
and it prevents the "bug" discussed in the comment above :)

tmolitor-stud-tu commented on schemas/interface.json:
```diff
@@ -2,12 +2,34 @@
+            },
+            "additionalProperties": true    // allow all sorts of entries (should be a json schema!)
+        },
+        "cmd_arguments_subschema": {
+            "type": "object",
+            "patternProperties": {
+                "^[a-zA-Z_][a-zA-Z0-9_.-]*$": { // arbitrary argument name
```
not really, at leas for the upcoming default arguments stuff already in my pipeline.
I figured having the subschemas all separated makes it easier to update them in the future and also seems a bit cleaner.

tmolitor-stud-tu commented:
I'm not sure if I can incorporate all these changes in time (e.g. in the next 1-2 weeks)

a-w50 commented:
This PR was merged via another branch
